crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -349,21 +401,15 @@ async fn half_bridge(
 
             // Forward messages from event loop to target
             Event::Incoming(Incoming::Publish(publish)) => {
+                received += 1;
                 if let Some(publish) = loop_breaker.ensure_not_looped(publish).await {
                     if let Some(topic) = transformer.convert_topic(&publish.topic) {
-                        target
-                            .publish(
-                                topic.clone(),
-                                publish.qos,
-                                publish.retain,
-                                publish.payload.clone(),
-                            )
-                            .await
-                            .unwrap();
+                        let duplicate = (topic.to_string(), publish.clone());
                         companion_bridge_half
-                            .send(Some((topic.into_owned(), publish)))
+                            .send((topic.to_string(), publish), Some(duplicate))
```

Comment:
It's a bit confusing seeing us send the same data twice to the same method. Now the bidirectional channel is no longer so general, maybe we should replace `send` with two methods, `forward_message` and `publish_novel_message` (or something along those lines). This is especially useful in this case as the reasoning behind the weird API is only apparent when looking at the healthcheck code, not the main loop.

Reply:
I can only agree that this is a really weird API!

Now that things are working, I will give a try to your proposal. An alternative is to send more specific structs.

Reply:
The duplicate is now created underneath by the publisher loop when required.

https://github.com/thin-edge/thin-edge.io/pull/3122/commits/3b2235f4d6ac4b369855047721b36b998b86b4eb